spikeextractors/extractors/neuroscopeextractors/neuroscopeextractors.py

```diff
@@ -0,0 +1,306 @@
+from spikeextractors import RecordingExtractor,SortingExtractor
+from spikeextractors.extractors.bindatrecordingextractor import BinDatRecordingExtractor
+import numpy as np
+from pathlib import Path
+from spikeextractors.extraction_tools import check_get_traces_args,check_valid_unit_id,read_binary
+from bs4 import BeautifulSoup
+import os
+
+try:
+    import bs4
+    import lxml
+    HAVE_BS4_LXML = True
+except ImportError:
+    HAVE_BS4_LXML = False
+
+class NeuroscopeRecordingExtractor(BinDatRecordingExtractor):
+    
+    """
+    Extracts raw neural recordings from large binary .dat files in the neuroscope format.
+    
+    The recording extractor always returns channel IDs starting from 0.
+    
+    The recording data will always be returned in the shape of (num_channels,num_frames).
+
+    Parameters
+    ----------
+    folder_path : str
+        Path to the folder containing the .dat file.
+    """
+    extractor_name = 'NeuroscopeRecordingExtractor'
+    installed = True  # check at class level if installed or not
+    is_writable = True
+    mode = 'folder'
+    installation_mesg = 'please install bs4 and lxml to use this extractor'  # error message when not installed
+            
+    def __init__(self, folder_path):
+        assert HAVE_BS4_LXML, self.installation_mesg
+        RecordingExtractor.__init__(self)
+        self._recording_file = folder_path
+        
+        fpath_base, fname = os.path.split(folder_path)
+        xml_filepath = os.path.join(folder_path, fname + '.xml')
+        dat_filepath = os.path.join(folder_path, fname + '.dat')
+        
+        with open(xml_filepath, 'r') as xml_file:
+            contents = xml_file.read()
+            soup = BeautifulSoup(contents, 'lxml')
+            # Normally, this would be a .xml, but there were strange issues
+            # in the write_recording method that require it to be a .lxml instead
+            # which also requires all capital letters to be removed from the tag names
+        
+        n_bits = int(soup.nbits.string)
+        dtype='int'+str(n_bits)
+        num_channels = int(soup.nchannels.string)
+        sampling_frequency = float(soup.samplingrate.string)
+        
+        BinDatRecordingExtractor.__init__(self, dat_filepath, sampling_frequency=sampling_frequency,
+                                          dtype=dtype, numchan=num_channels)
+        
+        self._kwargs = {'folder_path': str(Path(folder_path).absolute())}
+        
+        
+    @staticmethod
+    def write_recording(recording, save_path):
+        """ Convert and save the recording extractor to Neuroscope format
+
+        parameters
+        ----------
+        recording: RecordingExtractor
+            The recording extractor to be converted and saved
+        save_path: str
+            Full path to desired target folder
+        """
+        RECORDING_NAME = save_path + '.dat'
```

Comment:
Appending `.dat` is not neede as it's already handled by the `BinDatRecordingExtractor.write_recording()`, which calls the [`write_to_binary_dat_format`](https://github.com/SpikeInterface/spikeextractors/blob/f0d2653f6e55788150d0fcba4242aee5c75b8988/spikeextractors/extraction_tools.py#L294-L298) function.

So you can remove lines 74 and 78 and just pass `save_path` directly to the `write_recording`

Reply:
Got it. Also, it seems as if I'm unable to personally alter the travis.yaml (see conflicting files) to include bs4 and lxml, is that something you have to do?

Reply:
So you should add this repo as a remote upstream locally:

`git remote add upstream https://github.com/SpikeInterface/spikeextractors.git`

Then you can locally run (after you have committed your changes):

`git merge upstream/master`

Some conflicts will be there, you can resolve them locally, re-commit, and push. Then they should be fixed :)

Reply:
The remote upstream is already setup, and claims to be up to date on merge attempt. The latest commit to the /catalystneuro/spikeextractor fork, NeuroscopeRecordingExtractor branch, directly altered the travis.yaml at that location to include bs4 and lxml. After that commit, the travis-ci won't run because it shows the file conflicting.

Reply:
But do you have conflicts locally? If so, you have to resolve them, add the files, recommit, and push. Usually I'm able to fix it from github, but it doesn't seem possible now. If that doesn't work I can try to push on the branch directly later.

Reply:
No conflicts locally, the message is the one at the bottom of the pull request page where you usually see the travis results.

Reply:
And that "Only those with write access to this repository can merge pull requests."

Reply:
@CodyCBakerPhD when I checkout youor fork and run `git merge upstream/master` I got a merging conflict to the `.travis.yml`. Unfortunately I don't have write permissions on your fork. Next time you make a PR, there should be a checkbox sayng something like: allow owner to make changes. If you check that box, then I would be able to push on the repo. I think this is also why I can't resolve the conflict here: I can't commit and push to your fork.